iroh-resolver/src/balanced_tree.rs

```diff
@@ -0,0 +1,388 @@
+use std::collections::VecDeque;
+
+use anyhow::Result;
+use async_stream::try_stream;
+use bytes::{Bytes, BytesMut};
+use cid::Cid;
+use futures::{Stream, StreamExt};
+
+use crate::unixfs::{dag_pb, unixfs_pb, DataType, Node, UnixfsNode};
+use crate::unixfs_builder::encode_unixfs_pb;
+
+// 3                          root
+// 2                                 nextparent0
+// 1             parent0              parent1    ... parent7
+// 0 [0, 1, 2, 3, 4, 5, 6, 7]      [0, ... ]
+//
+
+pub fn stream_balanced_tree(
+    in_stream: impl Stream<Item = std::io::Result<BytesMut>>,
+    max_degrees: usize,
+) -> impl Stream<Item = Result<(Cid, Bytes)>> {
+    try_stream! {
+        // vec![ vec![] ]
+        // ..
+        // vec![ vec![0, 1, 2, 3, 4, 5, 6, 7] ]
+        // vec![ vec![8], vec![p0] ]
+
+        // ...
+
+        // vec![ vec![0, 1, 2, 3, 4, 5, 6, 7] vec![p0] ]
+        // vec![ vec![], vec![p0, p1]]
+
+        // ..
+
+        // vec![ vec![0, 1, 2, 3, 4, 5, 6, 7] vec![p0, p1, p2, p3, p4, p5, p6, p7], ]
+        // vec![ vec![], vec![p0, p1, p2, p3, p4, p5, p6, p7], vec![] ]
+        // vec![ vec![8], vec![p8], vec![pp0] ]
+
+        let mut tree: VecDeque<Vec<(Cid, u64)>> = VecDeque::new();
```

Comment:
Can you add a comment on what exactly we store in here? (mostly for ourselves in the future)